File: 3794-find-the-minimum-amount-of-time-to-brew-potions/3794-find-the-minimum-amount-of-time-to-brew-potions.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    long long minTime(std::vector<int>& skill, std::vector<int>& mana) {
        int n = static_cast<int>(skill.size());
        int m = static_cast<int>(mana.size());

        std::vector<long long> prev(n, 0);                 // prev column (T[·][j‑1])

        for (int j = 0; j < m; ++j) {
            std::vector<long long> temp(n, 0);             // current column (T[·][j])

            // ---------- left‑to‑right sweep ----------
            for (int i = 0; i < n; ++i) {
                long long left = (i == 0) ? 0               : temp[i - 1];
                long long up   = prev[i];
                temp[i] = 1LL * skill[i] * mana[j] + std::max(left, up);
            }

            // ---------- right‑to‑left “reverse” sweep ----------
            long long last = 1LL * skill[n - 1] * mana[j];
            for (int k = n - 2; k >= 0; --k) {
                temp[k] = std::max(temp[k], temp[k + 1] - last);
                last    = 1LL * skill[k] * mana[j];
            }

            prev.swap(temp);                               // roll column
        }

        return prev.back();                                // T[n‑1][m‑1]
    }
};

File: 3794-find-the-minimum-amount-of-time-to-brew-potions/3794-find-the-minimum-amount-of-time-to-brew-potions.py
class Solution:
    def minTime(self, skill: List[int], mana: List[int]) -> int:
        n: int = len(skill) # no of wizards
        m: int = len(mana) # no of potions
        f = [0] * (n + 1)

        
        for x in mana:
            for i in range(n):
                f[i + 1] = max(f[i + 1], f[i]) + (skill[i] * x)
            for i in reversed(range(n)):
                f[i] = f[i + 1] - (skill[i] * x)
            
        return f[n]

        